### PR TITLE
feat: implement fallback recommendations

### PR DESCRIPTION
**Pull request by zainab-amir**
### Description

If we get back <4 recommendations from Algolia, we should fill in the remaining spots (1, 2, 3, or all 4) with generic recommendations like we have as a fallback elsewhere, based on popularity/enrollment. 

```
GENERAL_RECOMMENDATIONS:
  - course_key: MITx+6.00.1x
    logo_image_url: https://prod-discovery.edx-cdn.org/organization/logos/2a73d2ce-c34a-4e08-8223-83bca9d2f01d-2cc8854c6fee.png
    marketing_url: https://www.edx.org/course/introduction-to-computer-science-and-programming-7
    title: Introduction to Computer Science and Programming Using Python
  - course_key: IBM+PY0101EN
    logo_image_url: https://prod-discovery.edx-cdn.org/organization/logos/87b07564-d569-4cfd-bee6-8b0a407acb73-dc33e4b5f353.png
    marketing_url: https://www.edx.org/course/python-basics-for-data-science
    title: Python Basics for Data Science
  - course_key: HarvardX+CS50P
    logo_image_url: https://prod-discovery.edx-cdn.org/organization/logos/44022f13-20df-4666-9111-cede3e5dc5b6-2cc39992c67a.png
    marketing_url: https://www.edx.org/course/cs50s-introduction-to-programming-with-python
    title: CS50's Introduction to Programming with Python
  - course_key: UQx+IELTSx
    logo_image_url: https://prod-discovery.edx-cdn.org/organization/logos/8554749f-b920-4d7f-8986-af6bb95290aa-f336c6a2ca11.png
    marketing_url: https://www.edx.org/course/ielts-academic-test-preparation
    title: IELTS Academic Test Preparation
  - course_key: HarvardX+CS50x
    logo_image_url: https://prod-discovery.edx-cdn.org/organization/logos/44022f13-20df-4666-9111-cede3e5dc5b6-2cc39992c67a.png
    marketing_url: https://www.edx.org/course/introduction-computer-science-harvardx-cs50x
    title: CS50's Introduction to Computer Science
```

#### How Has This Been Tested?

Added unit tests

#### Merge Checklist

* [ ] Is there adequate test coverage for your changes?

### CODE REVIEW

File: src/recommendations/RecommendationsPage.jsx [hunk 1]
```diff
@@ -77,7 +89,7 @@ const RecommendationsPage = (props) => {
           <Image className="logo" alt={getConfig().SITE_NAME} src={getConfig().LOGO_URL} />
         </Hyperlink>
       </div>
-      {(!isLoading && recommendations.length === RECOMMENDATIONS_COUNT) ? (
+      {(!isLoading && recommendations.length >= RECOMMENDATIONS_COUNT) ? (
```
Comment (syedsajjadkazmii) [inline src/recommendations/RecommendationsPage.jsx:92]: why did you change operator from `===` to `>=`?

File: src/recommendations/track.js [hunk 2]
```diff
@@ -16,11 +16,12 @@ export const createLinkTracker = (tracker, href, openInNewTab = false) => (e) =>
   return setTimeout(() => { global.location.href = href; }, LINK_TIMEOUT);
 };
 
-export const trackRecommendationsClicked = (courseKey, isControl, position, userId, href) => {
+export const trackRecommendationsClicked = (courseKey, isControl, position, userId, href, recommendationType) => {
   createLinkTracker(
     sendTrackEvent(eventNames.recommendedCourseClicked, {
       page: 'authn_recommendations',
       position,
+      recommendationType,
```
Comment (syedsajjadkazmii) [inline src/recommendations/track.js:24]: it should be 
```suggestion
      recommendation_type: recommendationType,
```

File: src/recommendations/RecommendationsPage.jsx [hunk 3]
```diff
@@ -29,14 +29,26 @@ const RecommendationsPage = (props) => {
...
+          const uniqueRecommendations = courseRecommendations.filter(
+            (recommendation, index, self) => index === self.findIndex((existingRecommendation) => (
+              existingRecommendation.courseKey === recommendation.courseKey
+            )),
+          );
+          setRecommendations(uniqueRecommendations.slice(0, RECOMMENDATIONS_COUNT));
         }
+
         setIsLoading(false);
+        // We only want to track the recommendations returned by Algolia
```
Comment (syedsajjadkazmii) [inline src/recommendations/RecommendationsPage.jsx:51]: does that mean we will be sending only those recommendations in track event which are returned by algolia?

Comment (zainab-amir) [reply src/recommendations/RecommendationsPage.jsx:51]: Yes. We want to keep track of what Algolia returned for a user so that we can see if those are the correct recs.

File: src/recommendations/RecommendationsPage.jsx (hunk 1 above)
Comment (zainab-amir) [reply src/recommendations/RecommendationsPage.jsx:92]: I'll revert this. I was doing some testing.

Comment (zainab-amir) [reply src/recommendations/RecommendationsPage.jsx:92]: But I think its okay even if the condition is `>=`. I'll keep it like this. It just means we won't show anything less than 4 but over 4 is okay.

Comment (syedsajjadkazmii) [reply src/recommendations/RecommendationsPage.jsx:92]: We are doing slicing to 4 recs. That means the recs can never be more than 4. I know this condition works but i think in code semantics, it does not make sense.

File: src/recommendations/RecommendationsPage.jsx (hunk 3 above)
Comment (syedsajjadkazmii) [reply src/recommendations/RecommendationsPage.jsx:51]: Then we should make the course keys array before concatenating the general recs. We need to move it up.

**Comment by zainab-amir** [comment]
@e0d
- I can use configs to get the `GENERAL_RECOMMENDATIONS` data.
- I am not sure what the protocol is to add code that is specific to and needed for edX but might not be useful for openedX community on frontends. Currently this code is behind a feature flag `ENABLE_PERSONALIZED_RECOMMENDATIONS`

**Review by syedsajjadkazmii — APPROVED** [review]

**Comment by nedbat** [comment]
Sorry if I'm not understanding.  The edX-specific course IDs look like they are in the test data only.  I'm not a front-end dev, so I could be mis-reading, and the PR description seems to indicate that these course IDs will be used as a default.

**Comment by arbrandes** [comment]
@nedbat, the mock data is actually not a huge deal, though we still don't think having hard-coded edx.org strings is ideal - even in test data.

The "algolia" fallback, on the other hand, represents more of an actual problem, even if hidden behind a feature flag.  We don't want to encourage adding features that can only be reasonably used by a single organization.

**Comment by brian-smith-tcril** [comment]
That being said, the [`getPersonalizedRecommendations` method itself](https://github.com/openedx/frontend-app-authn/blob/c31c03f5a9ead7ecf90c7074a93b63a03b025b98/src/recommendations/data/service.js) is tightly coupled to `algolia`

**Comment by nedbat** [comment]
> @nedbat, the mock data is actually not a huge deal, though we still don't think having hard-coded edx.org strings is ideal - even in test data.

Sure, but that's easy to fix: use truly fake data for the test data.

**Comment by zainab-amir** [comment]
@nedbat @brian-smith-tcril @e0d I am going to create a ticket in the backlog that will update the code to make it loosely coupled with Algolia. 
We'll work on it as a team to find a approach that works for everyone. Please note that it might be a few weeks before we can pick work to refactor it.
